Adds in-memory PCD serialization/deserialization; de-duplicates PCDReader::readHeader().

mattgruenke opened the pull request:
For my project, I need to transmit a series of binary, compressed PCD files as a stream.  In order to reduce latency and avoid potential wear on storage media, I want to serialize & deserialize them in-memory, rather than writing & reading them as files.

These changes enable this, as well as cleaning up a bit of mess I encountered along the way.  The biggest was removing a copy of PCDReader::readHeader() that differed in no ways justifying the fact that it didn't simply call the original.  I did compare both implementations using `diff` and review the history of this change, in order to reach this conclusion.

Any other cleanup was incidental, however I feel the file locking & mmap'ing is now easier to follow when a level of the call stack is devoted just to that.

Finally, I'd note that further cleanup & fixes of this code is possible, primarily by using RAII for file locking and mmap'ing.  I can only assume the principal author(s) weren't familiar with the technique, since it would make the code much safer, more readable, and more concise.  If there's interest, I'm willing to undertake this cleanup, but I think it'd be worthwhile getting my current change set merged in the interim.

I have tested my change by reading & writing PCD files, with and without additional in-memory serialization/deserialization in between.  The resulting files match the originals.

taketwo commented on io/include/pcl/io/pcd_io.h:
```diff
@@ -84,6 +84,32 @@ namespace pcl
+        * read/write PCD methods will detect column major input and automatically convert it.
+        *
+        * \param[in] binary_istream a std::istream with openmode set to std::ios::binary.
+        * \param[out] cloud the resultant point cloud dataset (only the header will be filled)
```
This is somewhat misleading. There is a member field called literally `header` inside the point cloud. But this function fills more than that.

mattgruenke replied:
It is functionally equivalent to the other readHeader() overloads.  I took this approach for the sake of maximizing consistency and minimizing duplication.  I can change it, if you'd like.

taketwo replied:
No, the approach is alright in my opinion. It's just the docstring that should be changed. Something in the way "only the meta-information fields will be filled".

mattgruenke commented:
Build job #3665.3 seems to have failed due to:

    The job exceeded the maximum time limit for jobs, and has been terminated.

The others succeeded, however.  Is this a common failure?  Nothing about my change should have a significant effect on build or run times.

taketwo commented on io/include/pcl/io/pcd_io.h:
```diff
@@ -168,6 +168,45 @@ namespace pcl
+        *  * == 0 on success
+        */
+      int
+      readBody_Ascii (std::istream &stream, pcl::PCLPointCloud2 &cloud, int pcd_version);
```
This does not adhere to PCL naming convention (no underscores).

mattgruenke replied:
Sorry.  What about readAsciiBody()?

taketwo replied:
There are already functions with names like `generateHeaderASCII` and `writeBinaryCompressed`, so I'd say the most consistent name will be `readBodyASCII`.

taketwo commented:
> The others succeeded, however. Is this a common failure? Nothing about my change should have a significant effect on build or run times.

No, it's likely not related. There is an ongoing work to solve the CI issue #1786.

mattgruenke commented on io/include/pcl/io/pcd_io.h:
```diff
@@ -252,12 +252,14 @@ namespace pcl
         */
-      std::string
-      generateHeaderBinaryCompressed (const pcl::PCLPointCloud2 &cloud,
+      int
```
My primary concerns were: consistency with the rest of the code in that file, and limiting the scope of my changes (which consist mostly of shuffling the existing code).  Much of this code is not how I'd have written it.  I'm happy to change anything you'd like (though I hope we can limit the scope, for this round).

So you want it to return void, and simply document that the state of the stream should be checked to determine success/failure?

taketwo replied:
I'd also implement this all differently. But you are right, we have to live with the rest of the code, so let's keep new things consistent. All functions here return (-1) for failure and (0) for success, let's keep it this way. But a note in docstring about this will still be helpful.

mattgruenke commented on io/include/pcl/io/pcd_io.h:
```diff
@@ -252,12 +252,14 @@ namespace pcl
-      std::string
-      generateHeaderBinaryCompressed (const pcl::PCLPointCloud2 &cloud,
+      int
+      generateHeaderBinaryCompressed (std::ostream &os,
```
Agreed.  I apologize for that oversight - I think I regraded it as a private member, but I see it's public.

mattgruenke commented:
I have tested on Linux (SuSE Linux Enterprise 12 SP2; x86_64).  I will soon test on Android NDK R14b API-21/ARM64 and possibly also Android API-19/ARMv7-A.  I can't easily test on Windows.

As mentioned above, this is mostly a reshuffling of the existing code.  The changes look more substantial than they really are.

mattgruenke commented:
BTW, thank you for reviewing my changes.

I should be able to implement your requested modifications, this weekend.  I'll also have a look at the unit tests.

UnaNancyOwen commented:
Thanks, This PR is a good feature.
I will check it on Windows when added unit tests.

taketwo commented:
@UnaNancyOwen :+1: thanks

mattgruenke commented:
BTW, it took me a while to find how to build & run the tests.  This helped:

https://github.com/PointCloudLibrary/pcl/issues/784

Even when BUILD_global_tests is enabled, I still found that the `tests` target was undefined.

SergioRAgostinho commented:
@mattgruenke You might want to consider rebase your PR to latest version of master. The unit tests suffered a massive refactoring, and depending on where and how you implement your unit tests, the rebase might prevent you from dealing with a merge conflict later on.

mattgruenke commented:
Apologies for the delays.

I've corrected some problems exposed by test_io.  I added a testcase for the new in-memory serialization.  These changes are now in my branch.

Next, I will rebase my branch.  BTW, my unit test changes are relatively isolated and minor.

taketwo commented on io/src/pcd_io.cpp:
```diff
@@ -1296,18 +1074,29 @@ pcl::PCDWriter::generateHeaderBinaryCompressed (const pcl::PCLPointCloud2 &cloud
+{
+  std::ostringstream oss;
+  generateHeaderBinaryCompressed (oss, cloud, origin, orientation);
+  return oss.str();
```
Missing space

taketwo commented on test/io/test_io.cpp:
```diff
@@ -1292,6 +1294,74 @@ TEST (PCL, LZFExtended)
+  EXPECT_EQ (blob2.height, blob.height);
+  EXPECT_EQ (data_type, 2); // since it was written by writeBinaryCompressed(), it should be compressed.
+
+  const unsigned char *data = reinterpret_cast<const unsigned char *> (pcd_str.data());
```
Missing space

mattgruenke commented:
So, do you need me to re-base?  Or is it good as-is?

taketwo commented:
I do not think a re-base is necessary as there are no conflicts. But the 14 commits of this PR is a bit too many, as I wrote some of them can be merged.

mattgruenke commented:
So, what would you like me to do?  As I understand it, the commits can be squashed into a single change.  Otherwise, what's the best option - re-create my fork and re-stage my changes as fewer commits?

I'm relatively inexperienced with git.

SergioRAgostinho commented:
Have a read https://www.atlassian.com/git/tutorials/rewriting-history  and use the interactive rebase to squash the commits which are logically/functionally related. 

By looking at your commit messages, it seems you did changes to PCLReader and to PCLWriter in separate commits. This could be for instance a logical division.

mattgruenke commented:
Thanks.

I think the bigger issue is that later commits fix formatting & errors in earlier ones, and they're not 1:1 related.

I was careful in how I staged my initial changes.  However, from that point onward, the process became fairly organic.

SergioRAgostinho commented:
Then just try cluster those organic changes into a commit with a message and description which expresses the overall changes.

taketwo commented:
We are just one step before merging this. @mattgruenke do you think you can fix the commit history as discussed? Would be a pity to leave out this otherwise well-crafted pull request.

mattgruenke commented:
I can try it, this weekend.

mattgruenke commented:
This is superseded by #1986, which I created by re-forking the repo and re-applying my changes in fewer commits (reduced from 13 to 6) that each follow the white space & naming conventions and avoid test failures.
